pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -181,101 +181,95 @@ func (s *DiscoveryServer) edsIncremental(version string, req *model.PushRequest)
-			ServiceAccounts: map[string]bool{},
-		}
-		s.EndpointShardsByService[serviceName][namespace] = ep
-		if !internal {
-			adsLog.Infof("Full push, new service %s", serviceName)
-			requireFull = true
-		}
+	fullPush := false
+
+	// Find endpoint shard for this service, if it is available - otherwise create a new one.
+	ep, created := s.findOrCreateEndpointShard(serviceName, namespace)
+
+	// If we create a new endpoint shard, that means we have not seen the service earlier. We should do a full push.
+	if created {
+		adsLog.Infof("Full push, new service %s", serviceName)
+		fullPush = true
 	}
 
-	// 2. Update data for the specific cluster. Each cluster gets independent
-	// updates containing the full list of endpoints for the service in that cluster.
-	serviceAccounts := map[string]bool{}
+	// Check if ServiceAccounts have changed. We should do a full push if they have changed.
+	serviceAccounts := sets.Set{}
 	for _, e := range istioEndpoints {
 		if e.ServiceAccount != "" {
-			serviceAccounts[e.ServiceAccount] = true
+			serviceAccounts.Insert(e.ServiceAccount)
 		}
 	}
 
-	if !reflect.DeepEqual(serviceAccounts, ep.ServiceAccounts) {
+	if !serviceAccounts.Equals(ep.ServiceAccounts) {
 		adsLog.Debugf("Updating service accounts now, svc %v, before service account %v, after %v",
 			serviceName, ep.ServiceAccounts, serviceAccounts)
-		if !internal {
-			requireFull = true
-			adsLog.Infof("Full push, service accounts changed, %v", serviceName)
-		}
+		adsLog.Infof("Full push, service accounts changed, %v", serviceName)
+		fullPush = true
 	}
 
 	ep.mutex.Lock()
 	ep.Shards[clusterID] = istioEndpoints
 	ep.ServiceAccounts = serviceAccounts
 	ep.mutex.Unlock()
 
-	// for internal update: this called by DiscoveryServer.Push --> UpdateServiceShards,
-	// no need to trigger push here.
-	// It is done in DiscoveryServer.Push --> AdsPushAll
-	if !internal {
-		s.ConfigUpdate(&model.PushRequest{
-			Full: requireFull,
-			ConfigsUpdated: map[model.ConfigKey]struct{}{{
-				Kind:      model.ServiceEntryKind,
-				Name:      serviceName,
-				Namespace: namespace,
-			}: {}},
-			Reason: []model.TriggerReason{model.EndpointUpdate},
-		})
+	return fullPush
+}
+
+func (s *DiscoveryServer) handleEmptyEndpoints(clusterID, serviceName, namespace string) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	// Should delete the service EndpointShards when endpoints become zero to prevent memory leak,
+	// but we should not do not delete the keys from EndpointShardsByService map - that will trigger
+	// unnecessary full push which can become a real problem if a pod is in crashloop and thus endpoints
+	// flip flopping between 1 and 0.
+	if s.EndpointShardsByService[serviceName][namespace] != nil {
+		s.deleteEndpointShards(clusterID, serviceName, namespace)
+		adsLog.Infof("Incremental push, service %s has no endpoints", serviceName)
+	}
+}
+
+func (s *DiscoveryServer) findOrCreateEndpointShard(serviceName, namespace string) (*EndpointShards, bool) {
```

Comment:
`getOrCreateEndpointShard` to keep consistent iirc

Reply:
ok. renamed